### PR TITLE
Some commands (in studio mainly) WIP

### PR DESCRIPTION
**Pull request by svdgoor**
PLEASE CHECK THE LAST COMMIT SPECIFICALLY, THANK!

### CODE REVIEW

**Comment by svdgoor** [comment]
**Change logs:**
Rename CMDIris<stuff> to DecIris<stuff>
Clean up & add function alternative to ProjectManager
Add /iris version
Create DecreeStudioExtension which:
- Extends DecreeExecutor with some checks that are used often

Add function to scriptExecutionEnvironment that allows for parsing an IrisScript instead of a script name
Add functions to DecreeExecutor:
- player() -> gets player
- message() -> messages player
- success() -> messages player in green
- warn() -> messages player in yellow
- error() -> messages player in red
- access() -> gets engine access in player's world
- world() -> gets players world

Replace required() on @Param and replace it with whether or not a default value was set
(Reason: If and only if there is a default, the parameter is not required. Hence, if `desc.isSet: req = false`)
Add [Iris]: tag to commands' results
Document sync() property on @Decree
Put JSONCleaner in a separate class in the util.JSON package
- I still need to convert your new command to use these functions

File: src/main/java/com/volmit/iris/Iris.java [hunk 1]
```diff
@@ -443,40 +443,38 @@ public void imsg(CommandSender s, String msg) {
 
     @Override
     public ChunkGenerator getDefaultWorldGenerator(String worldName, String id) {
-        String dimension = IrisSettings.get().getGenerator().getDefaultWorldType();
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/Iris.java]: Remove all changes from iris, this isnt related to commands, and this change was already merged in.

File: src/main/java/com/volmit/iris/core/CommandManager.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 package com.volmit.iris.core;
 
 import com.volmit.iris.Iris;
-import com.volmit.iris.core.decrees.CMDIris;
+import com.volmit.iris.core.decrees.DecIris;
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/core/CommandManager.java:22]: Justify this

File: src/main/java/com/volmit/iris/core/ProjectManager.java [hunk 3]
```diff
@@ -367,6 +363,10 @@ public void close() {
         }
     }
 
+    public File compilePackage(VolmitSender sender, IrisDimension d, boolean obfuscate, boolean minify) {
+        return new IrisProject(new File(getWorkspaceFolder(), d.getLoadKey())).compilePackage(sender, obfuscate, minify);
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/core/ProjectManager.java]: Dont do this, it will add complexity to the back end. SImply use the parameter `dimension.getLoadKey()` using the old method, dont duplicate methods to change a single param in these classes

File: src/main/java/com/volmit/iris/core/decrees/DecIris.java [hunk 4]
```diff
@@ -18,20 +18,26 @@
...
 
 @Decree(name = "irisd", aliases = {"ird"}, description = "Basic Command")
-public class CMDIris implements DecreeExecutor
+public class DecIris implements DecreeExecutor
 {
-    private CMDIrisStudio studio;
+    private DecIrisStudio studio;
 
-    @Decree(description = "Ping self", aliases = "p")
+    @Decree(description = "Send a message to yourself", aliases = "p")
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/core/decrees/DecIris.java]: This is a test command, it can actually be removed

File: src/main/java/com/volmit/iris/core/decrees/DecIris.java [hunk 5]
```diff
@@ -18,20 +18,26 @@
...
     public void ping(
-            @Param(name = "message",defaultValue = "Pong", aliases = {"msg", "m"})
+            @Param(name = "message", description = "The message to send", defaultValue = "Pong", aliases = {"msg", "m"})
             String message)
     {
         sender().sendMessage(message + "!");
     }
+
+    @Decree(description = "Print version information", aliases = {"v", "ver"})
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/core/decrees/DecIris.java]: Remove both of those aliases, Decree will recognize them already from 'version'

File: src/main/java/com/volmit/iris/core/decrees/DecreeStudioExtension.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.volmit.iris.core.decrees;
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/core/decrees/DecreeStudioExtension.java]: Put this in DecStudio, dont make extra classes just for studio.

File: src/main/java/com/volmit/iris/engine/IrisExecutionEnvironment.java [hunk 7]
```diff
@@ -54,9 +55,13 @@ public IrisScriptingAPI getAPI() {
     }
 
     public void execute(String script) {
-        Iris.debug("Execute Script (void) " + C.DARK_GREEN + script);
+        execute(getEngine().getData().getScriptLoader().load(script));
+    }
+
+    public void execute(IrisScript script) {
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/engine/IrisExecutionEnvironment.java:61]: Again, use script.getLoadKey, dont duplicate methods to reduce code in the command itself. Remove this method and use the old one with your command parameter

File: src/main/java/com/volmit/iris/engine/scripting/EngineExecutionEnvironment.java [hunk 8]
```diff
@@ -30,6 +31,8 @@
 
     void execute(String script);
 
+    void execute(IrisScript script);
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/engine/scripting/EngineExecutionEnvironment.java]: remove

File: src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java [hunk 9]
```diff
@@ -21,14 +21,45 @@
...
 public interface DecreeExecutor {
     default VolmitSender sender()
     {
         return DecreeContext.get();
     }
 
+    default Player player()
+    {
+        if (sender().isPlayer()) {
+            return sender().player();
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java]: Dont null check here. A NPE will fire either way if its not a player. Simply return sender.player() it makes the stack trace easier to read

File: src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java [hunk 10]
```diff
@@ -21,14 +21,45 @@
...
+        }
+        return null;
+    }
+
+    default void message(String string)
+    {
+        sender().sendMessage(string);
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java]: Dont do this, its not hard to type sender().message() in commands. Its harder to maintain say if we supported multiple parameters, your just adding technical debt here. Add this stuff to the sender, not the executor. Remove all message sending methods from this decree executor and add it to the sender itself

File: src/main/java/com/volmit/iris/util/decree/DecreeParameter.java [hunk 11]
```diff
@@ -57,7 +56,7 @@ public String getDescription() {
     }
 
     public boolean isRequired() {
-        return param.required();
+        return !hasDefault();
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/DecreeParameter.java:59]: I changed this for a reason, why change it back? We need default AND required. Yes i know they collide with each other but it doesnt cause any issues, where as this solution, removing required and ASSUMING required if the default isnt there, does not cover all edge cases. 

Revert this and any changes to the Param class too

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java [hunk 12]
```diff
@@ -28,15 +28,15 @@
...
+    String REQUIRED = "Required";
+
     /**
      * The main name of this command.<br>
      * Required parameter.<br>
      * This is what is used in game, alongside any (if specified) {@link #aliases() aliases}
      */
     String name();
 
-    boolean required() default false;
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/annotations/Param.java:38]: I changed this for a reason, why change it back? We need default AND required. Yes i know they collide with each other but it doesnt cause any issues, where as this solution, removing required and ASSUMING required if the default isnt there, does not cover all edge cases.

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java [hunk 13]
```diff
@@ -28,15 +28,15 @@
 public @interface Param {
     String DEFAULT_DESCRIPTION = "No Description Provided";
 
+    String REQUIRED = "Required";
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/annotations/Param.java]: stop

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java [hunk 14]
```diff
@@ -49,7 +49,7 @@
      * Which indicates the variable MUST be defined by the person running the command.<br>
      * If you define this, the variable automatically becomes non-required, but can still be set.
      */
-    String defaultValue() default "";
+    String defaultValue() default REQUIRED;
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/annotations/Param.java]: i explained this in vc

File: src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java [hunk 15]
```diff
@@ -411,7 +411,7 @@ public boolean invoke(VolmitSender sender, KList<String> args, KList<Integer> sk
         if(isStudio() && !IrisSettings.get().isStudio())
         {
             sender.sendMessage(C.RED + "To use Iris Studio Commands, please enable studio in Iris/settings.json (settings auto-reload)");
-            return false;
+            return true; // Must still return true because command exists but is not enabled
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java]: RETURN FALSE, THATS NOT WHAT FALSE OR TRUE MEANS

Return false so the tree searcher can continue to look for other commands that may match the input

File: src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java [hunk 16]
```diff
@@ -462,7 +462,7 @@ private boolean invokeNode(VolmitSender sender, KMap<String, Object> map) {
 
             try
             {
-                if(value == null && !i.getParam().defaultValue().trim().isEmpty())
+                if(value == null && i.hasDefault())
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java:465]: see above, no.

File: src/main/java/com/volmit/iris/util/json/JSONCleaner.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
+package com.volmit.iris.util.json;
```
Comment (cyberpwnn) [inline src/main/java/com/volmit/iris/util/json/JSONCleaner.java]: This should go where its used, in the studio commands

**Review by cyberpwnn — CHANGES_REQUESTED** [review]
I tried to help

File: src/main/java/com/volmit/iris/core/CommandManager.java (hunk 2 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/CommandManager.java:22]: Clarify the system is using Decree more than CMD
Also abbreviations shouldn't be all caps imo because of style.
Decree is our replacement for commands.

File: src/main/java/com/volmit/iris/core/ProjectManager.java (hunk 3 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/ProjectManager.java]: One

File: src/main/java/com/volmit/iris/core/decrees/DecIris.java (hunk 4 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/decrees/DecIris.java]: Two

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java (hunk 13 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/annotations/Param.java]: Three

File: src/main/java/com/volmit/iris/core/decrees/DecreeStudioExtension.java (hunk 6 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/decrees/DecreeStudioExtension.java]: Four

File: src/main/java/com/volmit/iris/engine/scripting/EngineExecutionEnvironment.java (hunk 8 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/engine/scripting/EngineExecutionEnvironment.java]: Why would we convert a string to an IrisScript, to a string, to an IrisScript and then run it?

File: src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java (hunk 9 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java]: Five

File: src/main/java/com/volmit/iris/util/decree/DecreeParameter.java (hunk 11 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/DecreeParameter.java:59]: Name me one case where either:
the absence of a default value means it's NOT a required parameter
OR
a required parameter has a default value

and I'll do it

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java (hunk 12 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/annotations/Param.java:38]: See above

File: src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java (hunk 15 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java]: Six

File: src/main/java/com/volmit/iris/util/json/JSONCleaner.java (hunk 17 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/json/JSONCleaner.java]: Question for this one, should I remove `#beautify` in favour of `#clean`?

**Review by svdgoor — COMMENTED** [review]
Patched half
Questions & discussion remain

File: src/main/java/com/volmit/iris/Iris.java (hunk 1 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/Iris.java]: I'll merge master into this again after #546 is resolved.
The only reason I added this here so it can actually be compiled and tested, Ill make sure not to include any changes compared to master in the Iris class near the end.

File: src/main/java/com/volmit/iris/engine/IrisExecutionEnvironment.java (hunk 7 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/engine/IrisExecutionEnvironment.java:61]: Seven

File: src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java (hunk 10 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/DecreeExecutor.java]: Does the same go for 
`#access`, `#world` and `#engine`?
Eight

File: src/main/java/com/volmit/iris/util/decree/annotations/Param.java (hunk 14 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/annotations/Param.java]: See above

Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/annotations/Param.java]: No actually this is resolved. It's using "" again with .isEmpty()

File: src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java (hunk 16 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/virtual/VirtualDecreeCommand.java:465]: Unsure why the hasDefault call here is wrong
hasDefault now does: `return !param.defaultValue().trim().isEmpty();`

File: src/main/java/com/volmit/iris/core/CommandManager.java (hunk 2 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/CommandManager.java:22]: tbh I would consider renaming them to DecreeIris<Sub> to match the current CommandIris<Sub>

File: src/main/java/com/volmit/iris/Iris.java (hunk 1 above)
Comment (cyberpwnn) [reply src/main/java/com/volmit/iris/Iris.java]: Merged just now so you can do that.

File: src/main/java/com/volmit/iris/core/CommandManager.java (hunk 2 above)
Comment (cyberpwnn) [reply src/main/java/com/volmit/iris/core/CommandManager.java:22]: Once we actually transfer lets call it CommandIris again, and CommandStudio, because they ARE commands, but for the time being i think dec should be fine.

File: src/main/java/com/volmit/iris/util/json/JSONCleaner.java (hunk 17 above)
Comment (cyberpwnn) [reply src/main/java/com/volmit/iris/util/json/JSONCleaner.java]: If you saw my clean command, it allowed you to add params like beautify rewriteObjects & update workspace all in the clean command. So yes. Idk what happened to my command but you could use that one since it uses a multicore multi-task job executor for it, looks super sexy ingame.

File: src/main/java/com/volmit/iris/util/decree/DecreeParameter.java (hunk 11 above)
Comment (cyberpwnn) [reply src/main/java/com/volmit/iris/util/decree/DecreeParameter.java:59]: If its required and has a default value, obviously Iris will still complain if you dont specify it. This case doesnt matter it works either way

However if it is NOT required and does NOT have a default value, NULL is inserted which is SUPER important. I actually use this in a few situations. For example. Iris pregen

/iris pregen [dim]

if dim is specified, parse & use it, otherwise (null, with NO DEFAULT AND NOT REQUIRED) will be used, in which case we can have custom logic to find what world they are currently in and assume that is their context. Im planning on using a commandContext to find out what engine they are talking about.

Basically, the default system is not as good as null checking and being smart about what null actually means. A default system cannot specify "my current location" for example.... nor would it be a good idea to make a system like that when we already have null lol.

**Comment by cyberpwnn** [comment]
Everything looks fine, however 2 things
* The required param still needs to come back
* Merge conflicts with iris

File: src/main/java/com/volmit/iris/Iris.java (hunk 1 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/Iris.java]: Merged remote tracking

File: src/main/java/com/volmit/iris/core/CommandManager.java (hunk 2 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/CommandManager.java:22]: 👌

File: src/main/java/com/volmit/iris/core/decrees/DecIris.java (hunk 5 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/core/decrees/DecIris.java]: Hot damn
Nine

File: src/main/java/com/volmit/iris/util/decree/DecreeParameter.java (hunk 11 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/decree/DecreeParameter.java:59]: This makes more sense than I'd like to admit :P

I'll get to it tomorrow, not enough focus to do it right today.

File: src/main/java/com/volmit/iris/util/json/JSONCleaner.java (hunk 17 above)
Comment (svdgoor) [reply src/main/java/com/volmit/iris/util/json/JSONCleaner.java]: OK
Ten

**Comment by svdgoor** [comment]
Todo:
- [Reintroduce required()](https://github.com/VolmitSoftware/Iris/pull/536#discussion_r689141451) or perhaps derivable (See Discord DMs @cyberpwnn)
- Test currently implemented commands before continuing
- Get answer on [this](https://github.com/VolmitSoftware/Iris/pull/536#discussion_r689137672) from @cyberpwnn 
- Get link to lines in code that allow parameter & command abbreviations (`/irisd version` also works as `/irisd v` & `ver`)
